File: src/scala/engine/movegen/NonSlidingMoveGenerator.scala
package scala.engine.movegen

import java.engine.board.bitboards.Bitboard
import scala.engine.board.bitboards.Bitboard.U64
import scala.engine.board.bitboards.Implicits._

/**
  * Created by melvic on 8/5/18.
  */
trait NonSlidingMoveGenerator extends BitboardMoveGenerator {
  type NonSlidingMove = (U64, U64, U64) => Long

  def allMoves: Stream[WithMoveType[NonSlidingMove]]

  def destinationBitsets: GeneratorStream[WithMoveType[U64]] = (bitboard, source, sideToMove) => {
    val pieces = Bitboard.singleBitset(source)
    val emptySquares = bitboard.emptySquares
    val opponents = bitboard.opponents(sideToMove)

    allMoves.map { case (move, moveType) => (move(pieces, emptySquares, opponents), moveType) }
  }
}

File: src/java/engine/movegen/bitboards/NonSlidingMoveGenerator.java
package java.engine.movegen.bitboards;

import java.engine.board.bitboards.Bitboard;
import java.engine.board.Piece;
import java.engine.utils.TriFunction;

import java.util.stream.Stream;

/**
 * Created by melvic on 7/31/18.
 */
public interface NonSlidingMoveGenerator extends BitboardMoveGenerator {
    @Override
    public default Stream<Long> generateAllMoves(Bitboard bitboard, int source, Piece.Side sideToMove) {
        long pieces = Bitboard.singleBitset(source);
        long emptySquares = bitboard.getEmptySquares();
        long opponents = bitboard.getSideBoard(sideToMove.opposite());

        return getAllMoves()
                .map(move -> move.apply(pieces, emptySquares, opponents))
                .filter(Bitboard::isNonEmptySet);
    }

    public Stream<TriFunction<Long, Long, Long, Long>> getAllMoves();
}
